### PR TITLE
Fix some calculation errors for ps_split_composite and make_polygon.

### PR DESCRIPTION
**Pull request by mephisto41**
There are some bugs when we have a transform-style:flat in the middle of transform-style:preserve-3d. So I fixed it by following changes.

1. ps_split_composite shader did not take position of dest_task into
account.
2. make_polygon didn't work if isolated_items_bounds is empty.

r? kvark

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1787)

### CODE REVIEW

File: webrender/src/frame_builder.rs [hunk 1]
```diff
@@ -46,12 +46,21 @@ fn make_polygon(
     stacking_context: &StackingContext,
     node: &ClipScrollNode,
     anchor: usize,
+    device_pixel_ratio: f32,
 ) -> Polygon<f32, WorldPixel> {
-    //TODO: only work with `isolated_items_bounds.size` worth of space
```
Comment (kvark) [inline webrender/src/frame_builder.rs]: why is this comment removed?

File: webrender/res/ps_split_composite.glsl [hunk 2]
```diff
@@ -38,11 +38,15 @@ void main(void) {
     CompositeInstance ci = fetch_composite_instance();
     SplitGeometry geometry = fetch_split_geometry(ci.user_data0);
     AlphaBatchTask src_task = fetch_alpha_batch_task(ci.src_task_index);
+    AlphaBatchTask dest_task = fetch_alpha_batch_task(ci.render_task_index);
+
+    vec2 dest_origin = dest_task.render_target_origin -
```
Comment (kvark) [inline webrender/res/ps_split_composite.glsl]: nice!

File: webrender/src/frame_builder.rs [hunk 3]
```diff
@@ -46,12 +46,21 @@ fn make_polygon(
...
     // from the primitive local coordinates into the layer transformation.
     // Which in turn needs it to be a render task property obeyed by all primitives
     // upon rendering, possibly not limited to `write_*_vertex` implementations.
+    if stacking_context.isolated_items_bounds.is_empty() {
+        // isolated_items_bounds is empty means there is no item for this sc.
+        // in this case, we use screen_bounds instead.
+        let rect = stacking_context.screen_bounds.to_f32() / device_pixel_ratio;
+        let bounds = LayerRect::new(LayerPoint::new(rect.origin.x, rect.origin.y),
+                                    LayerSize::new(rect.size.width, rect.size.height));
+        return Polygon::from_transformed_rect(bounds, LayerToWorldTransform::identity(), anchor);
```
Comment (kvark) [inline webrender/src/frame_builder.rs]: hmm, I'm not seeing the point here. If a stacking context has no items to isolate, why would we even want it to participate in the plane splitting (and hence, generate a polygon here)?

File: webrender/src/frame_builder.rs (hunk 1 above)
Comment (mephisto41) [reply webrender/src/frame_builder.rs]: I removed this accidentally. I'll add it back.

File: webrender/src/frame_builder.rs (hunk 3 above)
Comment (mephisto41) [reply webrender/src/frame_builder.rs]: In the reftest we have following display list.
```
sc1 (preserve3d)
  sc2 (preserve3d)
    rect1
  sc3 (flat)
    sc4 (preserve3d)
      sc5 (preserve3d)
        rect2
      sc6 (preserve3d)
        rect3
```

We'll create plane splitting for rect 2 and rect 3, then render it in sc3. Next, we'll create plane splitting for sc3 and render it on the screen. Thus, we have a sc with no item bound but need a plane splitting. Does it makes sense?

Comment (kvark) [reply webrender/src/frame_builder.rs]: I believe the `isolated_items_bounds` for SC3 would then need to include the bounds of SC4. Given that SC4 is transformed, its bounds would need to be projected onto SC3 (transformed).

**Comment by mephisto41** [comment]
@kvark , Now if the SC doesn't have any primitive I'll use the union of all children's bounds as isolated_items_bounds. Can you review it again? Thanks.

**Comment by kvark** [comment]
@mephisto41 looks good to me now, thank you!
Do you mind launching a try push to see if any Gecko issues appear before we proceed?

**Comment by mephisto41** [comment]
Sure, I'll submit a try.

**Comment by mephisto41** [comment]
Here is try.  https://treeherder.mozilla.org/#/jobs?repo=try&revision=9f31a21048f83634bed711b15e2a6b558e894805

**Comment by kvark** [comment]
@mephisto41 looks good, thank you!
Please rebase and r=me

**Comment by mephisto41** [comment]
@kvark, I found the original solution did not resolved the issue completely. If the SC is not isolation and it contains primitives and preserve-3d SCs. The bounds for make_polygon should include primitives bounds and all SC children bounds. My original solution only works with the non-isolation SC without any primitives. So I add a new test case for this problem and add a condition in the ```handle_pop_stacking_context```. I think this modification need your review.

```
else if stacking_context.isolation != ContextIsolation::Items {
    stacking_context.isolated_items_bounds = stacking_context
        .isolated_items_bounds
        .union(&stacking_context.children_sc_bounds);
}
```

**Comment by kvark** [comment]
> If the SC is not isolation and it contains primitives and preserve-3d SCs

Could you have a test case for that? AFAIS, both your reftests have all the contexts being isolated.

**Comment by mephisto41** [comment]
I think the SC in line 19 of intermediate-2.yaml is non-isolated SC which has a primitive and a preserve-3d SC.

**Comment by kvark** [comment]
@mephisto41 ok, I think we are good to go then, thank you!
@bors-servo r+

**Pull request merged**